### PR TITLE
Detect borrow checker errors where `.clone()` would be an appropriate user action

### LINKED ISSUE #49693: "error[E0507]: cannot move out of borrowed content" could suggest adding the Copy trait
For this error:

> error[E0507]: cannot move out of borrowed content
>    --> file.rs:302:35
>     |
> 302 |         let c = b.map(|entry| entry.symbol_kind);
>     |                                   ^^^^^ cannot move out of borrowed content
> 
> error: aborting due to previous error

The compiler could suggest adding #[derive(Copy, Clone)] to the relevant enum. The error message on its own does not offer enough insight on how to fix the problem, from a non-proficient user's POV.

### PR DESCRIPTION
**Pull request by estebank**
When a value is moved twice, suggest cloning the earlier move:

```
error[E0509]: cannot move out of type `U2`, which implements the `Drop` trait
  --> $DIR/union-move.rs:49:18
   |
LL |         move_out(x.f1_nocopy);
   |                  ^^^^^^^^^^^
   |                  |
   |                  cannot move out of here
   |                  move occurs because `x.f1_nocopy` has type `ManuallyDrop<RefCell<i32>>`, which does not implement the `Copy` trait
   |
help: consider cloning the value if the performance cost is acceptable
   |
LL |         move_out(x.f1_nocopy.clone());
   |                             ++++++++
```

When a value is borrowed by an `fn` call, consider if cloning the result of the call would be reasonable, and suggest cloning that, instead of the argument:

```
error[E0505]: cannot move out of `a` because it is borrowed
  --> $DIR/variance-issue-20533.rs:53:14
   |
LL |         let a = AffineU32(1);
   |             - binding `a` declared here
LL |         let x = bat(&a);
   |                     -- borrow of `a` occurs here
LL |         drop(a);
   |              ^ move out of `a` occurs here
LL |         drop(x);
   |              - borrow later used here
   |
help: consider cloning the value if the performance cost is acceptable
   |
LL |         let x = bat(&a).clone();
   |                        ++++++++
```

otherwise, suggest cloning the argument:

```
error[E0505]: cannot move out of `a` because it is borrowed
  --> $DIR/variance-issue-20533.rs:59:14
   |
LL |         let a = ClonableAffineU32(1);
   |             - binding `a` declared here
LL |         let x = foo(&a);
   |                     -- borrow of `a` occurs here
LL |         drop(a);
   |              ^ move out of `a` occurs here
LL |         drop(x);
   |              - borrow later used here
   |
help: consider cloning the value if the performance cost is acceptable
   |
LL -         let x = foo(&a);
LL +         let x = foo(a.clone());
   |
```

This suggestion doesn't attempt to square out the types between what's cloned and what the `fn` expects, to allow the user to make a determination on whether to change the `fn` call or `fn` definition themselves.

Special case move errors caused by `FnOnce`:

```
error[E0382]: use of moved value: `blk`
  --> $DIR/once-cant-call-twice-on-heap.rs:8:5
   |
LL | fn foo<F:FnOnce()>(blk: F) {
   |                    --- move occurs because `blk` has type `F`, which does not implement the `Copy` trait
LL |     blk();
   |     ----- `blk` moved due to this call
LL |     blk();
   |     ^^^ value used here after move
   |
note: `FnOnce` closures can only be called once
  --> $DIR/once-cant-call-twice-on-heap.rs:6:10
   |
LL | fn foo<F:FnOnce()>(blk: F) {
   |          ^^^^^^^^ `F` is made to be an `FnOnce` closure here
LL |     blk();
   |     ----- this value implements `FnOnce`, which causes it to be moved when called
```

Account for redundant `.clone()` calls in resulting suggestions:

```
error[E0507]: cannot move out of dereference of `S`
  --> $DIR/needs-clone-through-deref.rs:15:18
   |
LL |         for _ in self.clone().into_iter() {}
   |                  ^^^^^^^^^^^^ ----------- value moved due to this method call
   |                  |
   |                  move occurs because value has type `Vec<usize>`, which does not implement the `Copy` trait
   |
note: `into_iter` takes ownership of the receiver `self`, which moves value
  --> $SRC_DIR/core/src/iter/traits/collect.rs:LL:COL
help: you can `clone` the value and consume it, but this might not be your desired behavior
   |
LL |         for _ in <Vec<usize> as Clone>::clone(&self).into_iter() {}
   |                  ++++++++++++++++++++++++++++++    ~
```

We use the presence of `&mut` values in a move error as a proxy for the user caring about side effects, so we don't emit a clone suggestion in that case:

```
error[E0505]: cannot move out of `s` because it is borrowed
  --> $DIR/borrowck-overloaded-index-move-index.rs:53:7
   |
LL |     let mut s = "hello".to_string();
   |         ----- binding `s` declared here
LL |     let rs = &mut s;
   |              ------ borrow of `s` occurs here
...
LL |     f[s] = 10;
   |       ^ move out of `s` occurs here
...
LL |     use_mut(rs);
   |             -- borrow later used here
```

We properly account for `foo += foo;` errors where we *don't* suggest `foo.clone() += foo;`, instead suggesting `foo += foo.clone();`.

---

Each commit can be reviewed in isolation. There are some "cleanup" commits, but kept them separate in order to show *why* specific changes were being made, and their effect on tests' output.

Fix #49693, CC #64167.

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 1]
```diff
@@ -977,6 +977,21 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         can_suggest_clone
     }
 
+    pub(crate) fn clone_on_reference(&self, expr: &hir::Expr<'_>) -> Option<Span> {
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1211]: please add a doc comment, I don't fully understand this function.

Comment (lcnr) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1211]: it's poinit is to detect `x.clone()` calls where the type of `x` matches the return type which is caused by an auto-ref?

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 2]
```diff
@@ -977,6 +977,21 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+        let typeck_results = self.infcx.tcx.typeck(self.mir_def_id());
+        if let hir::ExprKind::MethodCall(segment, rcvr, args, span) = expr.kind
+            && let Some(expr_ty) = typeck_results.node_type_opt(expr.hir_id)
+            && let Some(rcvr_ty) = typeck_results.node_type_opt(rcvr.hir_id)
+            && rcvr_ty == expr_ty
+            && segment.ident.name == sym::clone
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1217]: is `clone` not a diagnostics item?

File: tests/ui/associated-types/associated-types-outlives.stderr [hunk 3]
```diff
@@ -10,6 +10,11 @@ LL |         drop(x);
    |              ^ move out of `x` occurs here
 LL |         return f(y);
    |                  - borrow later used here
+   |
+help: consider cloning the value if the performance cost is acceptable
+   |
+LL |         's: loop { y = denormalise(&x.clone()); break }
```
Comment (lcnr) [inline tests/ui/associated-types/associated-types-outlives.stderr]: that suggestion does not work as `x.clone()` gets dropped after the assignment 🤔

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 4]
```diff
@@ -1543,22 +1572,32 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
-            }
+            self.suggest_adding_bounds_or_derive(err, ty, clone, body.span);
+        }
+    }
+
+    pub(crate) fn suggest_adding_bounds_or_derive(
+        &self,
+        err: &mut Diag<'_>,
+        ty: Ty<'tcx>,
+        def_id: DefId,
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1812]:
```suggestion
        trait_def_id: DefId,
```

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 5]
```diff
@@ -1543,22 +1572,32 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+        span: Span,
+    ) {
+        self.suggest_adding_bounds(err, ty, def_id, span);
+        if let ty::Adt(..) = ty.kind() {
+            // The type doesn't implement DefId.
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1817]:
```suggestion
            // The type doesn't implement the trait.
```

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 6]
```diff
@@ -1024,27 +1024,32 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
                 ".clone()".to_owned()
             };
-        if let Some(clone_trait_def) = tcx.lang_items().clone_trait()
-            && self
-                .infcx
-                .type_implements_trait(clone_trait_def, [ty], self.param_env)
-                .must_apply_modulo_regions()
+        let mut sugg = Vec::with_capacity(2);
+        let mut inner_expr = expr;
+        while let hir::ExprKind::AddrOf(.., inner) | hir::ExprKind::Unary(hir::UnOp::Deref, inner) =
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1247]:
```rust
// Remove uses of `&` and `*` when suggesting `.clone()`.
```

File: tests/ui/associated-types/associated-types-outlives.rs [hunk 7]
```diff
@@ -3,10 +3,10 @@
 // fn body, causing this (invalid) code to be accepted.
 
 pub trait Foo<'a> {
-    type Bar;
+    type Bar: Clone;
```
Comment (lcnr) [inline tests/ui/associated-types/associated-types-outlives.rs:6]: why this change?

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 8]
```diff
@@ -1006,7 +1018,6 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                         None
                     }
                 })
-            && { true }
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs]: what was that doing here xx

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 9]
```diff
@@ -1048,7 +1051,58 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                             is_loop_message,
                         },
                     );
-                    err.subdiagnostic(self.dcx(), CaptureReasonNote::FnOnceMoveInCall { var_span });
+                    if let ty::Param(param_ty) = self_ty.kind()
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:1054]: please add a comment explaining what this is doing

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 10]
```diff
@@ -989,13 +989,109 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         can_suggest_clone
     }
 
+    fn suggest_cloning_on_spread_operator(
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:992]: `spread_operator`? `functional_record_update`

also: please cmment what this function is doing

File: tests/ui/borrowck/borrowck-struct-update-with-dtor.stderr [hunk 11]
```diff
@@ -1,26 +1,191 @@
...
+error[E0509]: cannot move out of type `S<K>`, which implements the `Drop` trait
+  --> $DIR/borrowck-struct-update-with-dtor.rs:29:19
+   |
+LL |         let _s2 = S { a: 2, ..s0 };
+   |                   ^^^^^^^^^^^^^^^^
+   |                   |
+   |                   cannot move out of here
+   |                   move occurs because `s0.c` has type `K`, which does not implement the `Copy` trait
+   |
+help: clone the value from the field instead of using the spread operator syntax
```
Comment (lcnr) [inline tests/ui/borrowck/borrowck-struct-update-with-dtor.stderr:86]: is spread operator syntax used anywhere else? I've only seen "functional record update/FRU" or simply "struct update syntax"

**Review by lcnr — COMMENTED** [review]
went through this PR and it mostly looks good.

I ended up only scimming over some of the functions as if let chains with 10 components are difficult to read.

nits, then r=me

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs (hunk 8 above)
Comment (estebank) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs]: That was me 1) trying to clean the order of commits into independent chunks and 2) trying to get the logic right 😅

File: tests/ui/borrowck/borrowck-struct-update-with-dtor.stderr (hunk 11 above)
Comment (estebank) [reply tests/ui/borrowck/borrowck-struct-update-with-dtor.stderr:86]: I think I've gotten confused with it being called that informally in internals, can certainly reword it.

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs (hunk 1 above)
Comment (estebank) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1211]: Yes, correct.

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs (hunk 2 above)
Comment (estebank) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1217]: It is, but the res in the method call is `Err` for these cases :-/

**Comment by estebank** [comment]
@bors r=lcnr

Addressed the comments, approving as per https://github.com/rust-lang/rust/pull/122603#pullrequestreview-1954525086

**Pull request merged**